[bug] Getting ERROR:HANDLER on five devices with

**First step**
- [x] I have read the FAQ https://github.com/seime/openhab-esphome?tab=readme-ov-file#faq

**My setup**
- [x] I'm running OH 5.1.4
- [x] The ESPHome device is running version 2026.7.2
- [x] Output of `openhab-cli` command `bundle:list|grep ESPHome` is:
`272 │ Active │  80 │ 4.1.0.202606050534    │ Seime Add-ons :: Bundles :: ESPHome Native API
`- [x] The output above is from the latest version obtained from https://github.com/seime/openhab-esphome/releases/tag/latest_oh4

**OH Logs**
- [x] Binding logs from logger `no.seime.openhab.binding.esphome` at level `DEBUG` is attached (HOWTO: https://www.openhab.org/docs/administration/logging.html#defining-what-to-log)
- [ ] If the issue is connectivity, I have included logs from binding startup to the issue has manifested itself

**Describe the bug**
Some things just don't get online, hanging on ERROR:HANDLER. This happened in the Marketplace version of the device as well as the latest Github version. Debug log:

```
2026-07-23 14:00:32.360 [DEBUG] [home.internal.handler.ESPHomeHandler] - [esphome:device:sfeerverlichting-vaas] Initializing ESPHome handler
2026-07-23 14:00:32.368 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.initialize()' on 'no.seime.openhab.binding.esphome.internal.handler.ESPHomeHandler@12b9ecc9': Task java.util.concurrent.FutureTask@56c03625[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@ed586f0[Wrapped task = no.seime.openhab.binding.esphome.internal.MonitoredCompositeExecutorService$TimedRunnable@74f4c17e]] rejected from java.util.concurrent.ThreadPoolExecutor@12c207f3[Running, pool size = 8, active threads = 8, queued tasks = 0, completed tasks = 1]
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@56c03625[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@ed586f0[Wrapped task = no.seime.openhab.binding.esphome.internal.MonitoredCompositeExecutorService$TimedRunnable@74f4c17e]] rejected from java.util.concurrent.ThreadPoolExecutor@12c207f3[Running, pool size = 8, active threads = 8, queued tasks = 0, completed tasks = 1]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2081) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:841) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1376) ~[?:?]
	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:123) ~[?:?]
	at no.seime.openhab.binding.esphome.internal.MonitoredCompositeExecutorService.schedule(MonitoredCompositeExecutorService.java:210) ~[?:?]
	at no.seime.openhab.binding.esphome.internal.handler.ESPHomeHandler.scheduleConnect(ESPHomeHandler.java:959) ~[?:?]
	at no.seime.openhab.binding.esphome.internal.handler.ESPHomeHandler.initialize(ESPHomeHandler.java:209) ~[?:?]
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:149) ~[?:?]
	at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1583) [?:?]
2026-07-23 14:00:32.386 [ERROR] [core.thing.internal.ThingManagerImpl] - Exception occurred while initializing handler of thing 'esphome:device:sfeerverlichting-vaas': Task java.util.concurrent.FutureTask@56c03625[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@ed586f0[Wrapped task = no.seime.openhab.binding.esphome.internal.MonitoredCompositeExecutorService$TimedRunnable@74f4c17e]] rejected from java.util.concurrent.ThreadPoolExecutor@12c207f3[Running, pool size = 8, active threads = 8, queued tasks = 0, completed tasks = 1]
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@56c03625[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@ed586f0[Wrapped task = no.seime.openhab.binding.esphome.internal.MonitoredCompositeExecutorService$TimedRunnable@74f4c17e]] rejected from java.util.concurrent.ThreadPoolExecutor@12c207f3[Running, pool size = 8, active threads = 8, queued tasks = 0, completed tasks = 1]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2081) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:841) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1376) ~[?:?]
	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:123) ~[?:?]
	at no.seime.openhab.binding.esphome.internal.MonitoredCompositeExecutorService.schedule(MonitoredCompositeExecutorService.java:210) ~[?:?]
	at no.seime.openhab.binding.esphome.internal.handler.ESPHomeHandler.scheduleConnect(ESPHomeHandler.java:959) ~[?:?]
	at no.seime.openhab.binding.esphome.internal.handler.ESPHomeHandler.initialize(ESPHomeHandler.java:209) ~[?:?]
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:149) ~[?:?]
	at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1583) [?:?]
```

--- a/src/main/java/no/seime/openhab/binding/esphome/internal/MonitoredCompositeExecutorService.java
+++ b/src/main/java/no/seime/openhab/binding/esphome/internal/MonitoredCompositeExecutorService.java
@@ -146,9 +146,8 @@ public class MonitoredCompositeExecutorService implements ScheduledExecutorServi
             return new FakeScheduledFuture<>(executor.submit(
                     new TimedRunnable(command, getStackTraceElements(), defaultMaxExecutionTimeMs, null, true)));
         }
-        return new CompondScheduledFuture<>(scheduler.schedule(
-                () -> executor.submit(
-                        new TimedRunnable(command, getStackTraceElements(), defaultMaxExecutionTimeMs, null, true)),
+        return new CompondScheduledFuture<>(scheduler.schedule(() -> submitOrLog(
+                new TimedRunnable(command, getStackTraceElements(), defaultMaxExecutionTimeMs, null, true), null),
                 delay, unit));
     }
 
@@ -157,7 +156,27 @@ public class MonitoredCompositeExecutorService implements ScheduledExecutorServi
         if (delay <= 0L) {
             return new FakeScheduledFuture<>(executor.submit(callable));
         }
-        return new CompondScheduledFuture<>(scheduler.schedule(() -> executor.submit(callable), delay, unit));
+        return new CompondScheduledFuture<>(scheduler.schedule(() -> submitOrLog(callable, null), delay, unit));
+    }
+
+    private <V> Future<V> submitOrLog(Callable<V> task, @Nullable String callerSignature) {
+        try {
+            return executor.submit(task);
+        } catch (RejectedExecutionException e) {
+            logger.warn("Task '{}' rejected by executor: {}", callerSignature != null ? callerSignature : "<unnamed>",
+                    e.getMessage());
+            throw e;
+        }
+    }
+
+    private Future<?> submitOrLog(Runnable task, @Nullable String callerSignature) {
+        try {
+            return executor.submit(task);
+        } catch (RejectedExecutionException e) {
+            logger.warn("Task '{}' rejected by executor: {}", callerSignature != null ? callerSignature : "<unnamed>",
+                    e.getMessage());
+            throw e;
+        }
     }
 
     @Override
@@ -207,12 +226,16 @@ public class MonitoredCompositeExecutorService implements ScheduledExecutorServi
     public @Nullable ScheduledFuture<?> schedule(Runnable command, int delay, TimeUnit timeUnit, String callerSignature,
             long maxExecutionTimeMs) {
         if (delay <= 0L) {
-            return new FakeScheduledFuture<>(executor.submit(
-                    new TimedRunnable(command, getStackTraceElements(), maxExecutionTimeMs, callerSignature, false)));
+            return new FakeScheduledFuture<>(submitOrLog(
+                    new TimedRunnable(command, getStackTraceElements(), maxExecutionTimeMs, callerSignature, false),
+                    callerSignature));
         }
-        return new CompondScheduledFuture<>(scheduler.schedule(() -> executor.submit(
-                new TimedRunnable(command, getStackTraceElements(), maxExecutionTimeMs, callerSignature, false)), delay,
-                timeUnit));
+        return new CompondScheduledFuture<>(
+                scheduler
+                        .schedule(
+                                () -> submitOrLog(new TimedRunnable(command, getStackTraceElements(),
+                                        maxExecutionTimeMs, callerSignature, false), callerSignature),
+                                delay, timeUnit));
     }
 
     private class FakeScheduledFuture<V> implements ScheduledFuture<V> {

--- a/src/main/java/no/seime/openhab/binding/esphome/internal/handler/ESPHomeHandlerFactory.java
+++ b/src/main/java/no/seime/openhab/binding/esphome/internal/handler/ESPHomeHandlerFactory.java
@@ -152,8 +152,8 @@ public class ESPHomeHandlerFactory extends BaseThingHandlerFactory {
                 return t;
             });
 
-            ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(4, maxPoolSize, 60L, TimeUnit.SECONDS,
-                    new SynchronousQueue<>(), r -> {
+            ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(maxPoolSize, maxPoolSize, 60L,
+                    TimeUnit.SECONDS, new LinkedBlockingQueue<>(1000), r -> {
                         long currentCount = threadCounter.incrementAndGet();
                         logger.debug("Creating new worker thread {} for scheduler", currentCount);
                         Thread t = new Thread(r);
@@ -161,6 +161,7 @@ public class ESPHomeHandlerFactory extends BaseThingHandlerFactory {
                         t.setName("ESPHome Thing Executor " + currentCount);
                         return t;
                     });
+            threadPoolExecutor.allowCoreThreadTimeOut(true);
 
             scheduler = new MonitoredCompositeExecutorService(scheduledExecutorService, threadPoolExecutor, 300);
 

